### ISSUE TITLE
Processing v3 proposal

### ISSUE THREAD

**Issue by jleclanche** [HearthSim/HSReplay.net#276]
Brainstorming a new processing infrastructure with no Lambdas for processing itself. We are happy with the isolated lambda and don't want to touch it.

### High level goals/requirements

- The final result must be significantly cheaper than our current Lambda architecture.
- We don't want to be burdened by managing a lot of infrastructure and micro-optimizing costs just yet, so we want it to be overprovisioned and still cheaper than current Lambdas.
- We don't want to increase the complexity of the processing infrastructure; if possible we want to decrease it.
- We want to retain our high resilience against failure; data loss is still our number one enemy. Whatever is handling processing and queueing, if it goes down, we must be able to requeue it all.
- We want to retain or increase our control over the queue itself: Pausing, unpausing, requeueing and managing throughput.
- We want it to be dead easy to scale vertically and horizontally.
- If processing fails to keep up, we don't want this to be catastrophic: Maybe replays just take longer to show up.
- We don't want to spend a month on this project. Things to do, places to be.

### Lessons learned from current infrastructure

- Kinesis is expensive, atrociously complex and super inflexible. Can we do without it?
- S3 calls are expensive. We should minimize them.
- Cloudwatch is **really expensive** and a stray print() call can cost us thousands.
- Using S3 as a processing queue is resilient but impractical: It's hard to count the amount of objects in S3, and turning on/off relies on cutting lambda triggers off. Requeuing is also very slow and we have little insight into the current state of the queue.
- The ability to pause, unpause, requeue, reprocess are all critical.
- Idempotent processing is important to avoid bad/lingering data, but should be done in the database. We are underutilizing postgres locking.
- Lambda throttling is a nasty potential data loss vector which we are super scared of.
- Slower processing times have a butterfly effect on our infrastructure: More db connections, slower reprocessing, etc. 
- If processing slows down because of the database, this snowballs into slower and slower lambdas, which end up timing out, getting throttled, etc.
- The Lambda containerized update system is really, *really* nice.
- Lambda canaries have given us a ton of confidence in our infrastructure.

### What can't we do right now?

Some stories that would be nice to have:

- What is the current backlog size in item count **and** estimated time? (= avg processing duration + exact item count in the queue, including for reprocessing)
- Please reprocess this replay that no longer has a log file (= just reprocess its metadata using the xml file)
- Can we take the opportunity to refactor the game db locking? Maybe move all database calls to the end of the processing?
- If there is an increased amount of errors in a short period, we want processing to automatically pause.
- We want processing to automatically pause when the db is down, unpause when it comes back up.
- We would like it to be easy to prioritize processing selectively (eg. based on auth token/user)
- Log to disk instead of to cloudwatch. Inspect logs from the disk, find them very quickly by shortid. /var/log/hsreplaynet-processing/shortids/aaaaaaaaaa.log (XXX: But when scaling horizontally, this becomes cumbersome due to lack of log centralization)

### Brainstorming

Rethinking about this almost from scratch.

So let's just go from the assumption that we want processing to happen on EC2 nodes. We can scale the amount of nodes up/down and make sure we pick the appropriate instance type for our CPU+Memory load (at the end of the day, we have 4 potential bottlenecks: CPU, database load, network connectivity, processing dispatcher load).

We'll go with a minimum of two concurrent EC2 instances to process log files, so that we don't get tempted to think about a single-instance infrastructure.

Say we retain the S3 "queue" system, but we no longer use it as our primary queue. Instead, we have a global Redis queue (we can start with a managed Redis from AWS, although again a non-managed one may be a lot cheaper depending on our usage), written to by the isolated lambda. If Redis goes down, we recover by scanning our s3 raw bucket and putting stuff back into Redis.

Similarly to Lambda, we want our Python runtime to be preloaded in memory and already have an open connection to postgres etc. The job would therefore be a long-running process, which constantly polls from Redis, pops an item from it, processes it. Once processing finishes, the program moves on to the next item. If the queue is empty, sleep 1 second or less and reattempt.

We can run multiple instances of this with eg. supervisor numprocs=4 and max out our CPU. We constantly feed metadata into influx as well to make sure we have a lot of insight into the process.

The redis queue would look like a fifo: shortid + added date timestamp. We append as we add to the queue, the processing job pops the first item in the queue. We may need to lock processing more strigently in the uploads database in order to ensure things can't get processed twice if they are already undergoing processing (eg. something was queued twice into redis for some reason).

If processing crashes, if redis crashes, if anything crashes, stuff is still present on S3 in the raw bucket and that part still works the same. We can even schedule a "requeue all raw uploads" every few hours.

I believe all of this would:

- Massively reduce processing costs (couple of EC2s to handle the load + a maybe-managed Redis instance).
- Greatly increase our control over processing (no more throttles, no more "we're stuck on a particular version of Python", no more psycopg2 compile weirdness, no more pain setting up VPCs, can now control timeout settings, etc...).
- Greatly increase insight into our processing system, as well as finer-grained control over the state of processing: Increase/decrease throughput by adding nodes/processes/manual throttles.
- Decrease the complexity of the reprocessing infrastructure by getting rid of Kinesis and using the same system for processing *and* reprocessing instead.
- Be a stepping stone to factoring out the Lambda code from the website code.

**Comment by jleclanche** [comment]
Following up on this after we talked about using SQS as processing queue, which is a less ambitious project than all this refactoring.

Let's reiterate the issues we have with S3 as the processing queue:

* It's expensive. API calls account for 75% of our S3 expenses.
* It's opaque. Getting metrics on our queue state is hard, we have to look at the raw bucket size which we have to calculate on the fly...
* It's complex. Really, we're reinventing the wheel here. We have to reimplement a lot of things that we get for free with a proper queue service.

At the time, we went for an S3 queue because it was a solid evolution of the system we had before -- in hindsight, we should have looked at the alternative of not using S3. Not doing so cost us a lot of money in the long run.

So let's look at refactoring *only the queue* rather than then Lambda system.

## What is our current queue?

This is the current pipeline:

1. User pings uploads.hsreplay.net API (isolated lambda)
2. API stores user's message in S3 (the "descriptor") and replies with an S3 PUT URL
3. User uploads to PUT URL
4. S3 event triggers on upload, launching the processing lambda
5. Processing lambda looks for expected descriptor, then reads both log + descriptor, deletes the descriptor once it's in the upload DB.

In the current architecture, the descriptors are our messages and the S3 events dictates what gets processed and when. If we pause processing, when we unpause, all the messages that have piled up have to get scanned and re-triggered using SNS.
If processing fails before the DB is reached, we have very little insight into that. Additionally, for reprocessing, we have to use Kinesis and SNS and it goes through a different lambda... it's all very confusing.

## How would this work under SQS?

So first of all, I'd like to see processing lambdas consolidated, instead of having multiple lambdas for the different "types" of processing which are actually all the same.
Second, I'd like to at least have the ability to move into self-hosted processing in the future, at least partially.

Now here's the thing to keep in mind: When the API request is made, we're not yet ready to process it. In fact, we might never be ready; it's only OK to process once the log file has been uploaded.
That means we can't queue things into SQS and then blindly poll them. This is kind of a problem because that's exactly how SQS operates. If we queue the descriptors into SQS, when we retrieve them we have to be ready to process them.
It's also worth remembering that we're using this complex processing queue system so that our uptime is always Amazon S3's uptime and we don't lose things when our DB is down. What if our DB could make high uptime guarantees, would it still make sense to do that?

Open questions:

* Where do descriptors live in the time between the API call and the upload?
* How do we design the system so that popping from the processing queue always yields a descriptor with a log ready to parse?
* How do we make sure that we're not adding a lot of latency to processing a single upload? (The HDT click-through experience is super nice right now; really fast)
* Do we keep our metadata separated from our data? Does that still make sense?
* How do we keep the system reactive rather than passive?
* Are the log files part of the queue or not? ([Fun relevant read](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-s3-messages.html))

The system would be a lot simpler if we could guarantee that every descriptor is accompanied with a log file; that means we can queue both into SQS together. But... at that point, we don't really need SQS anymore in regular operation; only if we want to throttle or pause the queue. Our system should have the 99% case as its primary concern, not the 1% case.

So broken down, we want volatile storage for the metadata awaiting the log file; then once we have it, we want both to be in a fifo-like processing queue.
That final queue is where we can either pop a single message to process it in a lambda, or long-poll messages in batches (if we want to process with our own fleet). So really, open question 1 is what needs answering: When the log file is uploaded, how do we fetch the matching descriptor?

If we keep storing the descriptor in S3 this problem kinda answers itself; we retain two uploads, but the raw bucket is no longer our queue. Instead, the log upload creates an SQS message, which is later consumed by a Lambda (and the Lambda will know how to find the matching descriptor the same way as it does now). But it doesn't feel quite right; the descriptors would be better off in SQS themselves wouldn't they?

**Comment by amw2104** [comment]
A couple of potentially relevant SQS features:

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-delay-queues.html

This might be useful for creating a window for the log to get uploaded before we attempt to process it. We can use an HDT influx metric to determine the average delay between sending the descriptor and finishing the PUT with the log data and then configure the queue to use that delay. The downside of this is that we would always pay this delay even when the log arrives faster than average. A slightly more sophisticated strategy would be to have each HDT installation keep track of its average delay locally, send that number along with the descriptor, and then have the isolated lambda override the default MessageVisibility delay for that descriptor, which would reduce the number of retry attempts from HDT users with slow internet connections.

http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ChangeMessageVisibility.html

If we pursue a strategy where we don't have a guarantee that the log has arrived by the time we first attempt to process the descriptor, then we might consider using ChangeMessageVisibility to implement some sort of exponential decay between retry attempts.

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html

Dead Letter Queues might be useful for us whenever we have bad log files or orphan descriptors so that our queue doesn't get clogged up and we have an easy way to review the logs that fail parsing.

**Comment by jleclanche** [comment]
> If we pursue a strategy where we don't have a guarantee that the log has arrived by the time we first attempt to process the descriptor, then we might consider using ChangeMessageVisibility to implement some sort of exponential decay between retry attempts.

I thought about that, but my concern (on top of always paying the delay) is that constantly attempting to pop from the queue will create API noise and be more expensive than it could be.

Regarding DLQs, they're definitely useful to handle any errors we get (which currently get re-routed inside the raw upload bucket itself), but I don't think we should use them for failed parsing as that's a type of validation error. By that point it's already in the DB, I think our current uploads db system is fine.

Here's what I'm working from:

1. Uploader pushes metadata to uploads.hsreplay.net and receives a PUT URL in return.
2. Metadata is stored in S3 as a descriptor. Descriptors have a 24 hour lifetime (expire after 1d).
3. Uploader pushes log file to the PUT URL
4. S3 event queues a SQS message
5. ??? something triggers a Lambda to process the SQS message
6. If the Lambda fails (eg. descriptor/log is missing, problem connecting to the db, etc) it goes into DLQ, is retried n times later
7. Otherwise everything else works exactly as before.

Basically what this does is it removes the concept of the bucket being the processing queue, and instead stuff is accumulated into SQS. It's not 100% ideal because it introduces one more weak link into the chain (sqs down, s3 events flaky, etc) which means we'll have data but no way to put it into the queue.

The other thing is, what triggers the Lambda? Can it trigger upon a SQS message being queued? Is that even a good idea? The more I think about it, the more I think this would be a lot simpler *without* the Lambda component; then with this system we could simply have a couple of long-polling servers which process messages as they become available.

**Comment by jleclanche** [comment]
In fact if we have continuous processes reading the SQS stream, then that solves the descriptor issue: They can live in SQS, as the order they arrive into the process does not matter.

Imagine this:

- Q1: Descriptor queue
- Q2: Replay queue
- P1: One (or more) long-polling process that reads the descriptor queue
- P2: One (or more) long-polling process that reads the replay queue

Both P1 and P2 input the data they read into the uploads database. When either process notices that both descriptor and file are present, the upload is handed off to the processing function (which can be a lambda).

This isn't super elegant; requires two queues and two separate managed processes. Can we do simpler?

**Comment by jleclanche** [comment]
Let's review from scratch. Here is the simplest design for a "replay processing function":

1. User pushes both data + metadata in the same request to an endpoint
2. The endpoint immediately processes the data and returns either an error or the finished replay.

Problems:

- Pushing out log data into an API request is awkwardly ugly (gotta use multipart/form-data)
- Log data makes the body very large (to the point that it's not supported by api gateway)
- The function has a very large scope and if it crashes, permanent data loss occurs

1 is not a big deal. We fixed 2 by moving the file upload directly to s3, which required splitting up the metadata and data pushes. We made the metadata push a very thin function, which fixed 3.

So the design at that point is:

1. User pushes metadata first, metadata is uploaded to S3 and user gets a PUT URL
2. User pushes data to PUT URL, triggers an S3 event
3. A second process picks up the S3 event to process both data and metadata.

Problems:

- We want the ability to pause replay processing and recover from that point on

This is where we introduced the "raw bucket" concept. Log files in the raw bucket are considered unprocessed; we pause processing by removing the s3 event trigger and stuff accumulates in the bucket. We recover by iterating over every file in the bucket.

So this is where we went wrong; S3 is not a good accumulator for this sort of thing and it's costing us in API requests.

Quick pricing analysis: Looking at a *complete descriptor* right now (with headers and everything), it's about 2.5kb. At the very conservative 4096 bytes, times the very-optimistic 60 million replays / month, storing *all* of that in SQS costs $22/month.
Requests to SQS are $0.40 per million. Even if we average out 8 requests per replay (which is way higher than what we could achieve since we can batch 10 replays per request), for 60 million replays / month, that's $192/month. 

Right now, our S3 requests are: PUT upload, PUT descriptor, GET descriptor, GET upload, COPY upload, DELETE upload, DELETE descriptor. $0.000005/req (PUT/COPY), $0.0000004/req (GET), that's ~$950/month for 60 million replays (I'm not counting the final replay PUT). We're higher than that for reasons...

**Comment by jleclanche** [comment]
In our most recent iteration, we figured Redis would be a good temporary hold for descriptors but that fell through because we need reliable persistence, which redis is not good at.

Now it's time to re-review introducing Postgres in the upload pipeline. Specifically, could we simply reintroduce the uploads database in the uploads lambda?

It would simplify a lot: If we keep a fall-through to S3, this is what the first three steps look like:

1. User pushes metadata to API gateway, this triggers uploads lambda
2. Uploads lambda saves metadata in directly in uploads db and returns a PUT URL
3. User uploads to PUT URL, triggering S3 event
  - (SQS branch) 
    1. S3 event is pushed to SQS
    2. Continuous gatekeeper absorbs SQS event, triggers a processing job with the shortid
  - (Current branch)
    1. S3 event triggers processing job
4. Processing job queries descriptor from db and log from s3, processes the upload.

The branching is very simple. In the case that the uploads db is down, this is what it looks like:

1. User pushes metadata to API gateway, this triggers uploads lambda
2. Uploads lambda attempts to save metadata in uploads db but times out; saves to s3 instead and returns a PUT URL
3. ...
4. Processing job queries descriptor from db, does not find it, looks for it in s3 instead. Finds it in s3, gets log from s3, processes the upload.

**Comment by jleclanche** [comment]
Update:
As of yesterday, we are now using RDS for temporary descriptor storage. This falls back to S3 in case RDS is not available. Descriptors older than 14 days are automatically deleted.

The last remaining piece of work is to move to SQS for event storage instead of doing immediate handling. Putting this on hold for a few days.

**Comment by jleclanche** [comment]
Doesn't look like the rest of the v3 will go through. Fine as is for now. @joolean you can have a look at all this if you get inspired and want to reopen later.
